### DISCUSSION TITLE
[general] Unexpected behaviour when detecting relationships between targets

### DISCUSSION THREAD

**Discussion by agila5** [ropensci/targets#1156] [General]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Dear all, today I faced a weird error when working on a project that uses `targets`. Reprex of the problem: 

``` r
library(targets)

tar_dir({
  tar_script(
    code = {
      list(
        tar_target(
          name = "x", 
          command = 1
        ), 
        tar_target(
          name = "y", 
          command = {
            x = x + 1
            x
          }
        )
      )
    }
  )
  tar_make()
})
#> ▶ start target x
#> ● built target x [0 seconds]
#> ▶ start target y
#> ✖ error target y
#> ▶ end pipeline [0.19 seconds]
#> Error:
#> ! Error running targets::tar_make()
#>   Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
#>   Debugging guide: https://books.ropensci.org/targets/debugging.html
#>   How to ask for help: https://books.ropensci.org/targets/help.html
#>   Last error: object 'x' not found
#> Backtrace:
#>     ▆
#>  1. └─targets::tar_make()
#>  2.   └─targets:::callr_outer(...)
#>  3.     ├─targets:::if_any(...)
#>  4.     └─targets:::callr_error(condition = out, fun = fun)
#>  5.       └─targets::tar_throw_run(message, class = class(condition))
#>  6.         └─targets::tar_error(...)
#>  7.           └─rlang::abort(message = message, class = class, call = tar_empty_envir)
```

<sup>Created on 2023-09-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Everything works fine if I remove the `x = x + 1` step or just write something like `x_new = x + 1` (suggesting that targets fails to recognise the relationship between the two targets when using a syntax like `x = x + 1`). I checked the docs online and I didn't find anything useful related to this error, so I'm not sure if this is really a bug or just expected behaviour or a known limitation. Anyway, thank you very much for maintaining this great package!

**Comment by wlandau** [comment]
Please see https://github.com/ropensci/targets/issues/453. This comes from the fact that `targets` relies on the `codetools` package (c.f. https://github.com/ropensci/targets/issues/453#issuecomment-834411214). Your workaround using `x_new = x + 1` is a good one. As explained in https://books.ropensci.org/targets/targets.html#dependencies, you may also find other problems with dependency relationships automatically inferred from R code, which is why it is recommended to check the `tar_visnetwork()`/`tar_glimpse()` dependency graph before running a new or changed pipeline.

**Reply by agila5** [reply]
Thank you very much for the explanation!
